### PR TITLE
Runtime rework

### PR DESCRIPTION
**Pull request by gctucker**
Rework the `Runtime` class to make it better aligned with the new API & Pipeline.

### CODE REVIEW

**Comment by gctucker** [comment]
Tested OK on staging with job submitted by hand:
https://bot.staging.kernelci.org/job/gtucker/job/gtucker-kernel-build-trigger/16/console
https://staging.kernelci.org/test/job/gtucker/branch/kernelci-staging/kernel/kernelci-staging-20230306.0/

**Comment by gctucker** [comment]
Tested OK on staging with the new API & Pipeline:
```
03/06/2023 01:43:35 PM UTC [INFO] 2023-03-06 13:43:35.785458  396702e0748e  6405edf0d32374a94665efdd  Done       Pass      kver
```

File: kernelci/runtime/__init__.py [hunk 1]
```diff
@@ -3,44 +3,68 @@
...
+def add_kci_raise(jinja2_env):
+    """Add a kci_raise function to use in templates
 
-def add_kci_raise(env):
+    This adds a `kci_raise` function to a given Jinja2 environment `jinja2_env`
+    so it can be used to raise exceptions from template files when for example
+    some template parameters are not valid.
+    """
     def template_exception(msg):
         raise Exception(msg)
```
Comment (mgalka) [inline kernelci/runtime/__init__.py:25]: It's not part of this PR, but maybe it'd be good to use specialized exception class here like `TemplateError` or `TemplateParameterError` instead of a generic `Exception`. This could improve further exceptions handling.

File: kernelci/runtime/kubernetes.py [hunk 2]
```diff
@@ -3,28 +3,28 @@
...
-        template_path = plan_config.get_template_path(None)
-        template = jinja2_env.get_template(template_path)
-        job_name = '-'.join([
-            'kci', params['node_id'], params['name'][:24]
-        ])
+    def generate(self, params, plan_config):
+        template = self._get_template(plan_config)
+        job_name = '-'.join(['kci', params['node_id'], params['name'][:24]])
         safe_name = re.sub(r'[\:/_+=]', '-', job_name).lower()
         rand_sx = ''.join(random.sample(self.RANDOM_CHARACTERS, 8))
```
Comment (mgalka) [inline kernelci/runtime/kubernetes.py:29]: Again, this change hasn't been introduced here, but maybe it's a good idea to change `RANDOM_CHARACTERS` name to something more descriptive - this characters are not really random. Maybe  simply use `(ascii_lowercase+digits)` here.

**Review by mgalka — COMMENTED** [review]
small suggestions

File: kernelci/runtime/kubernetes.py (hunk 2 above)
Comment (gctucker) [reply kernelci/runtime/kubernetes.py:29]: I think `RANDOM_CHARACTERS` is actually better as the set is an implementation detail.  We could call it `RANDOM_K8S_JOB_NAME_CHARACTERS` but that's longer and not very useful for a class that's not used anywhere else.

File: kernelci/runtime/__init__.py (hunk 1 above)
Comment (gctucker) [reply kernelci/runtime/__init__.py:25]: Yes, error handling needs to be revisited throughout the code base in general.  In this case, we might need to define several handlers or add a variable with the type of exception to raise.  Also there might be something more suitable in recent versions of Jinja2.

Comment (mgalka) [reply kernelci/runtime/__init__.py:25]: Maybe we could make it a follow-up?

File: kernelci/runtime/kubernetes.py (hunk 2 above)
Comment (gctucker) [reply kernelci/runtime/kubernetes.py:29]: OK so something like `K8S_JOB_NAME_CHARACTERS` would be more descriptive, even though I think that's still not the best name we can come up with.

Comment (gctucker) [reply kernelci/runtime/kubernetes.py]: Alright I've renamed it.

File: kernelci/runtime/__init__.py (hunk 1 above)
Comment (gctucker) [reply kernelci/runtime/__init__.py:25]: #1797

**Comment by gctucker** [comment]
@mgalka I think I've addressed all the points, unless there's anything else I guess this could be merged later today after the production update.

**Comment by mgalka** [comment]
> @mgalka I think I've addressed all the points, unless there's anything else I guess this could be merged later today after the production update.

All looks good.

**Review by mgalka — APPROVED** [review]

**Comment by gctucker** [comment]
Or before :p There was no rush as this isn't required in production, but as long as it's not breaking anything then fine.
